src/livestatus.c

```diff
@@ -0,0 +1,479 @@
+      i++;
+
+      if (!exit) {
+        if (i >= LIVESTATUS_EXPECTED_FIELDS_RESP_NB) {
+          ERROR("livestatus plugin: too many fields in livestatus output");
+          rc = -2;
+          goto free_all_fields;
+        }
+      }
+    }
+
+    if (i < LIVESTATUS_EXPECTED_FIELDS_RESP_NB) {
+      ERROR("livestatus plugin: not enough fields in livestatus output");
+      rc = -3;
+      goto free_all_fields;
+    }
+
+    *lstatus = c_to_livestatus_status((const char **)fields);
+
+    rc = 0;
+
+  free_all_fields:
+    for (int j = 0; j < LIVESTATUS_EXPECTED_FIELDS_RESP_NB; j++) {
+      if (fields[j] != NULL) {
+        free(fields[j]);
+        fields[j] = NULL;
+      }
+    }
+
+    free(fields);
+    fields = NULL;
+
+    return rc;
+  } /* int ls_parse */
+
+  static int ls_read_parse(const int sockfd, livestatus_status_t *lstatus) {
+    ssize_t bread = -1;
+    char buffer[IO_BUFFER_SIZE];
+
+    do {
+      memset(buffer, 0x0, sizeof(buffer));
+      bread = read((int)sockfd, buffer, IO_BUFFER_SIZE - 1);
+    } while (errno == EINTR);
+
+    if (bread < 0) {
+      ERROR("livestatus plugin: reading from socket: %s", STRERRNO);
+      return -1;
+    }
+
+    buffer[IO_BUFFER_SIZE - 1] = '\0';
+
+    return ls_parse((const char *)buffer, lstatus);
+  } /* int ls_read_parse */
+
+  static int ls_send_request(const int sockfd) {
+    int written = -1;
+    char request[512];
+
+    memset(request, 0x0, sizeof(request));
+
+    snprintf(request, sizeof(request) - 1, "GET status\nColumns: %s\n\n",
+             LIVESTATUS_QUERY_COLUMNS);
+
+    do {
+      written = write(sockfd, request, sizeof(request));
+    } while (errno == EINTR);
+
+    if (written < 0) {
+      ERROR("livestatus plugin: writing to socket: %s", STRERRNO);
+      return -1;
+    }
+
+    if (shutdown(sockfd, SHUT_WR) < 0) {
+      ERROR("livestatus plugin: closing socket write: %s", STRERRNO);
+      return -2;
+    }
+
+    return 0;
+  } /* int ls_send_request */
+
+  static value_list_t ls_collectd_init_vl(void) {
+    value_list_t vl = VALUE_LIST_INIT;
+
+    vl.values_len = 1;
+    sstrncpy(vl.plugin, LIVESTATUS_PLUGIN_NAME, sizeof(vl.plugin));
+
+    return vl;
+  } /* value_list_t ls_collectd_init_vl */
+
+  static int ls_collectd_dispatch_count(gauge_t value,
+                                        const char *plugin_instance) {
+    value_list_t vl = ls_collectd_init_vl();
+
+    sstrncpy(vl.type, "count", sizeof(vl.type));
+    sstrncpy(vl.plugin_instance, (char *)plugin_instance,
+             sizeof(vl.plugin_instance));
+    vl.values = &(value_t){.gauge = value};
+
+    return plugin_dispatch_values(&vl);
+  } /* int ls_collectd_dispatch_count */
+
+  static int ls_collectd_dispatch_counter(counter_t value,
+                                          const char *plugin_instance) {
+    value_list_t vl = ls_collectd_init_vl();
+
+    sstrncpy(vl.type, "counter", sizeof(vl.type));
+    sstrncpy(vl.plugin_instance, (char *)plugin_instance,
```

Comment:
The cast to `char *` is not necessary – the second argument to `sstrncpy` is a `const char *`.

Reply:
Fixed.